### PR TITLE
Add VoteManager and endpoint to manage voting (VoteServlet)

### PR DESCRIPTION
**Pull request by limli**

### CODE REVIEW

File: src/main/java/com/google/step/datamanager/VoteManagerDatastore.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    // If entity exists, use it. Else, create a new entity.
+    if (entity == null) {
+      entity = new Entity("Vote");
+      entity.setProperty("user", userId);
+      entity.setProperty("deal", dealId);
+    }
+
+    entity.setProperty("dir", dir);
+    datastore.put(entity);
```
Comment (jyx11011) [inline src/main/java/com/google/step/datamanager/VoteManagerDatastore.java:53]: would you like to support undoing upvote/downvote?

File: src/main/java/com/google/step/datamanager/VoteManager.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.google.step.datamanager;
+
+public interface VoteManager {
+  /** Gets the number of votes the Deal with the given dealId has */
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/VoteManager.java]: nit: grammar

"Gets the number of votes for a deal, specified by dealId."

File: src/main/java/com/google/step/datamanager/VoteManager.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  public int getVotes(long dealId);
+
+  /**
+   * The user with the given userId votes on the Deal with the given dealId in the given direction.
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/VoteManager.java]:
```
Registers a vote from a user (specified by {@code userId}) for a deal (specified by {@code dealId}) in {@code dir} direction.
```

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]
Please add tests for VoteManagerDatastore.

File: src/main/java/com/google/step/datamanager/VoteManager.java [hunk 4]
```diff
@@ -1,11 +1,12 @@
 package com.google.step.datamanager;
 
 public interface VoteManager {
-  /** Gets the number of votes the Deal with the given dealId has */
+  /** Gets the number of votes for a deal, specified by dealId. */
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/VoteManager.java]:
```suggestion
  /** Gets the number of votes for a deal, specified by {@code dealId}. */
```

File: src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+@RunWith(JUnit4.class)
+public final class VoteManagerDatastoreTest {
+
+  private static final long USER_ID_A = 123;
+  private static final long USER_ID_B = 456;
+  private static final long USER_ID_C = 789;
+
+  private static final long DEAL_ID_A = 111;
+  private static final long DEAL_ID_B = 222;
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java]: Is the plan to eventually move to the common constants being defined in #27?

File: src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void testVote() {
+    // // original should have 0 votes
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java]:
```suggestion
    // original should have 0 votes
```

File: src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void testVote() {
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java]: This test is actually testing a lot of different scenarios. Please split it up into the following:

- `getVotes_returnsZeroAtFirst()`
- `vote_incrementsVotesCorrectly()`
- `vote_doesNotIncrementForDuplicateVoteFromUser()`
- `vote_doesNotAffectOtherDeals()`

Comment (SuyashLakhotia) [reply src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java]: Also, add additional tests to ensure the direction parameter is working properly. Perhaps:

* `vote_decrementsVotesCorrectly()`
* `vote_doesChangeVoteForSameUser()` (if the same user votes again with a different direction)

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java (hunk 5 above)
Comment (limli) [reply src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java]: yes

File: src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java [hunk 8]
```diff
@@ -36,30 +35,50 @@ public void tearDown() {
...
     manager.vote(USER_ID_A, DEAL_ID_A, 1);
-    assertEquals(manager.getVotes(DEAL_ID_A), 2);
+    assertEquals(manager.getVotes(DEAL_ID_A), 1);
+    manager.vote(USER_ID_A, DEAL_ID_A, 1);
+    assertEquals(manager.getVotes(DEAL_ID_A), 1);
+  }
 
-    // vote count should decrease when someone votes -1
+  @Test
+  public void vote_decrementVotesCorrectly() {
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java]:
```suggestion
  public void vote_decrementsVotesCorrectly() {
```

File: src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java [hunk 9]
```diff
@@ -36,30 +35,50 @@ public void tearDown() {
...
+  @Test
+  public void vote_doesNotAffectOtherDeals() {
+    manager.vote(USER_ID_A, DEAL_ID_A, 1);
+    manager.vote(USER_ID_A, DEAL_ID_B, -1);
+    assertEquals(manager.getVotes(DEAL_ID_A), 1);
     assertEquals(manager.getVotes(DEAL_ID_B), -1);
   }
 
+  @Test
+  public void vote_userChangeDirection() {
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/VoteManagerDatastoreTest.java:75]: Let's add another test where the vote should be removed? `vote_userUndoesVote()`?

```
manager.vote(USER_ID_A, DEAL_ID_A, 1);
assertEquals(manager.getVotes(DEAL_ID_A), 1);
manager.vote(USER_ID_A, DEAL_ID_A, 0);
assertEquals(manager.getVotes(DEAL_ID_A), 0);
```

File: src/main/java/com/google/step/datamanager/VoteManagerDatastore.java (hunk 1 above)
Comment (SuyashLakhotia) [reply src/main/java/com/google/step/datamanager/VoteManagerDatastore.java:53]: +1, IIUC this is supported by `dir == 0`?

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/step/datamanager/VoteManagerDatastore.java (hunk 1 above)
Comment (limli) [reply src/main/java/com/google/step/datamanager/VoteManagerDatastore.java:53]: yes `dir == 0` should undo

**Review by SuyashLakhotia — DISMISSED** [review]

**Review by SuyashLakhotia — DISMISSED** [review]

**Review by SuyashLakhotia — APPROVED** [review]

**Pull request merged**